src/core/include/mp-units/bits/fixed_point.h

```diff
@@ -0,0 +1,309 @@
+  Tl lo;
+};
+
+#if defined(__SIZEOF_INT128__)
+MP_UNITS_DIAGNOSTIC_PUSH
+MP_UNITS_DIAGNOSTIC_IGNORE("-Wpedantic")
+using int128_t = __int128;
+using uint128_t = unsigned __int128;
+MP_UNITS_DIAGNOSTIC_POP
+inline constexpr std::size_t max_native_width = 128;
+#else
+using int128_t = double_width_int<std::int64_t>;
+using uint128_t = double_width_int<std::uint64_t>;
+constexpr std::size_t max_native_width = 64;
+#endif
+
+template<typename T>
+constexpr std::size_t integer_rep_width_v = std::numeric_limits<std::make_unsigned_t<T>>::digits;
+template<typename T>
+constexpr std::size_t integer_rep_width_v<double_width_int<T>> = double_width_int<T>::width;
+
+template<typename T>
+constexpr bool is_signed_v = std::is_signed_v<T>;
+template<typename T>
+constexpr bool is_signed_v<double_width_int<T>> = double_width_int<T>::is_signed;
+
+template<typename T>
+using make_signed_t =
+  std::conditional_t<!std::is_same_v<T, uint128_t>, std::make_signed<T>, std::type_identity<int128_t>>::type;
+
+template<std::size_t N>
+using min_width_uint_t =
+  std::tuple_element_t<std::max<std::size_t>(4u, std::bit_width(N) + (std::has_single_bit(N) ? 0u : 1u)) - 4u,
+                       std::tuple<std::uint8_t, std::uint16_t, std::uint32_t, std::uint64_t, uint128_t>>;
+
+template<std::size_t N>
+using min_width_int_t = make_signed_t<min_width_uint_t<N>>;
+
+template<typename T>
+using double_width_int_for_t = std::conditional_t<is_signed_v<T>, min_width_int_t<integer_rep_width_v<T> * 2>,
+                                                  min_width_uint_t<integer_rep_width_v<T> * 2>>;
+
+template<typename Lhs, typename Rhs>
+constexpr auto wide_product_of(Lhs lhs, Rhs rhs)
+{
+  if constexpr (integer_rep_width_v<Lhs> + integer_rep_width_v<Rhs> <= max_native_width) {
+    using T = std::common_type_t<double_width_int_for_t<Lhs>, double_width_int_for_t<Rhs>>;
+    return static_cast<T>(lhs) * static_cast<T>(rhs);
+  } else {
+    using T = double_width_int<std::common_type_t<Lhs, Rhs>>;
+    return T::wide_product_of(lhs, rhs);
+  }
+}
+
+// This class represents rational numbers using a fixed-point representation, with a symmetric number of digits (bits)
+// on either side of the decimal point. The template argument `T` specifies the range of the integral part,
+// thus this class uses twice as many bits as the provided type, but is able to precisely store exactly all integers
+// from the declared type, as well as efficiently describe all rational factors that can be applied to that type
+// and neither always cause underflow or overflow.
+template<std::integral T>
+struct fixed_point {
+  using repr_t = double_width_int_for_t<T>;
+  static constexpr std::size_t fractional_bits = integer_rep_width_v<T>;
+
+  constexpr fixed_point() = default;
+  constexpr fixed_point(const fixed_point&) = default;
+
+  constexpr fixed_point& operator=(const fixed_point&) = default;
+
+  explicit constexpr fixed_point(repr_t v) : int_repr_is_an_implementation_detail_(v) {}
```

Comment:
Let's use `consteval` here as well.

I understand that in other libraries this might be useful as a runtime-enabled type. However, this is my primary concern here. External libraries might start to depend on our implementation details because of this. This will make us harder to change the bits if needed without breaking user's code. We would be unable to remove it from the repo if we possibly find a better solution, or change its implementation (ABI breaks).

Reply:
Ok, that is a valid concern. Users should rather use a dedicated fixed-point library such as [CNL](https://github.com/johnmcfarlane/cnl).

There is still the issue of testing though, I fear about the compilation time if we are going to test a significant number of values and value combinations.

Reply:
I do not think it will be worse than the thousands of tests we have for quantities already. But if some members are needed at runtime, then it makes sense to keep this type of runtime-enabled.